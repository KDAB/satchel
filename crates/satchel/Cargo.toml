[package]
name = "satchel"
version = "0.1.1"
license = "MIT OR Apache-2.0"
repository = "https://github.com/KDAB/satchel/"
edition = "2024"
description = "Test registration and discovery crate. Satchel uses linkme for test registration and exposes APIs for test discovery."

[dependencies]
libtest-mimic = "0.8"
linkme = "0.3"
satchel-macro = "0.1.0"

