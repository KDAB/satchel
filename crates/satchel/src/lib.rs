#![no_std]

pub use satchel_macro::{bench, test};

pub type TestFn = fn();

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum TestKind {
    Unit,
    Benchmark,
}

#[derive(Debug, Clone)]
pub struct ShouldPanic {
    pub expected: Option<&'static str>,
}

#[derive(Debug, Clone)]
pub struct Ignore {
    pub reason: Option<&'static str>,
}

#[derive(Debug, Clone)]
pub struct TestCase {
    pub name: &'static str,
    pub module_path: &'static str,
    pub kind: TestKind,
    pub test_fn: TestFn,
    pub should_panic: Option<ShouldPanic>,
    pub ignore: Option<Ignore>,
}

pub mod test_harness {
    pub use crate::{Ignore, ShouldPanic, TestCase};
    use linkme::distributed_slice;

    #[doc(hidden)]
    #[distributed_slice]
    pub static TESTS: [TestCase];
}

#[doc(hidden)]
pub fn extract_crate_name(module_path: &str) -> &str {
    module_path
        .split("::")
        .next()
        .expect("Split never returns a empty iterator")
}

#[doc(hidden)]
pub fn get_tests_for_crate(crate_prefix: &str) -> impl Iterator<Item = &'static TestCase> {
    let crate_name = extract_crate_name(crate_prefix);
    test_harness::TESTS
        .iter()
        .filter(move |case| case.module_path.starts_with(crate_name))
}

#[macro_export]
macro_rules! get_tests {
    () => {
        ::satchel::get_tests_for_crate(::core::module_path!())
    };
}

#[cfg(test)]
mod tests {
    use super::extract_crate_name;

    #[test]
    fn handles_empty_string() {
        assert_eq!(extract_crate_name(""), "");
    }

    #[test]
    fn extracts_simple_crate_name() {
        assert_eq!(extract_crate_name("mycrate"), "mycrate");
    }

    #[test]
    fn extracts_crate_name_with_module() {
        assert_eq!(extract_crate_name("mycrate::foo"), "mycrate");
        assert_eq!(extract_crate_name("mycrate::foo::bar"), "mycrate");
    }
}
