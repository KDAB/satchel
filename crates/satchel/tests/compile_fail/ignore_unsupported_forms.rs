mod common;
use satchel::test;

// Invalid form with parentheses and identifier
#[test]
#[ignore(bad)]
fn bad_ignore_form() {}

// Positional literal not supported
#[test]
#[ignore("reason")]
fn ignored_positional() {}

// Non-string literal reason
#[test]
#[ignore = 123]
fn ignored_non_string() {}

// Duplicate ignore attributes
#[test]
#[ignore]
#[ignore = "reason"]
fn duplicate_ignore() {}

fn main() {}