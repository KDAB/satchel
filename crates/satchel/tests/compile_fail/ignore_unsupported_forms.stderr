error: only #[ignore] and #[ignore = "..."] forms are supported
 --> tests/compile_fail/ignore_unsupported_forms.rs:6:1
  |
6 | #[ignore(bad)]
  | ^^^^^^^^^^^^^^

error: only #[ignore] and #[ignore = "..."] forms are supported
  --> tests/compile_fail/ignore_unsupported_forms.rs:11:1
   |
11 | #[ignore("reason")]
   | ^^^^^^^^^^^^^^^^^^^

error: only #[ignore] and #[ignore = "..."] forms are supported
  --> tests/compile_fail/ignore_unsupported_forms.rs:16:1
   |
16 | #[ignore = 123]
   | ^^^^^^^^^^^^^^^

error: duplicate #[ignore] attribute
  --> tests/compile_fail/ignore_unsupported_forms.rs:22:1
   |
22 | #[ignore = "reason"]
   | ^^^^^^^^^^^^^^^^^^^^
